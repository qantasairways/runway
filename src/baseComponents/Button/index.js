import React from 'react';
import PropTypes from 'prop-types';
import { withTheme } from 'emotion-theming';

import { CSS_SELECTOR_ACTIVE } from '../../constants/css';
import noop from '../../utils/noop';
import deepMerge from '../../utils/deepMerge';

const baseStyles = {
  button: (theme, props) => ({
    fontFamily: theme.fontFamily.main,
    fontSize: theme.fontSize.button,
    width: props.inline ? 'auto' : '100%',
    height: '3.125rem',
    borderRadius: theme.layout.borderRadius,
    borderWidth: 0,
    padding: '14px 24px 15px',
    fontWeight: 700,
    lineHeight: 1.56,
    letterSpacing: '0.1em',
    textTransform: 'uppercase',
    cursor: 'pointer',
    [CSS_SELECTOR_ACTIVE]: {
      transform: `translate(1px, 1px)`
    }
  })
  // Nested compoenets would go here for example
  // icon: theme => ({})
};

export function ThemedButton({ children, theme, themeStyles, ...buttonProps }) {
  return (
    <button
      type="button"
      {...buttonProps}
      css={deepMerge(
        baseStyles.button(theme, buttonProps),
        themeStyles.button(theme, buttonProps)
      )}
    >
      {children}
    </button>
  );
}

export const buttonPropTypes = {
  children: PropTypes.node,
  type: PropTypes.string,
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  inline: PropTypes.bool
};

export const buttonDefaultProps = {
  children: 'button',
  type: 'button',
  onClick: noop,
  disabled: false,
  inline: false
};

ThemedButton.propTypes = {
  theme: PropTypes.shape().isRequired,
  themeStyles: PropTypes.shape({ button: PropTypes.func }),
  ...buttonPropTypes
};

ThemedButton.defaultProps = {
  children: 'button',
  themeStyles: { button: noop },
  ...buttonDefaultProps
};

export default withTheme(ThemedButton);
