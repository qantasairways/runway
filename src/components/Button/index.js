import { css } from '@emotion/core';
import React from 'react';
import PropTypes from 'prop-types';

import { CSS_SELECTOR_ACTIVE, CSS_SELECTOR_HOVER } from '../../constants/css';
import { fontFamily, fontSize, colours, buttons } from '../../theme/airways';

export const BUTTON_VARIANT_SOLID = 'solid';
export const BUTTON_VARIANT_HOLLOW = 'hollow';

const baseStyles = css({
  fontFamily: fontFamily.main,
  fontSize: fontSize.button,
  width: buttons.width,
  height: buttons.height,
  borderRadius: buttons.borderRadius,
  borderWidth: 0,
  padding: '14px 24px 15px',
  fontWeight: 700,
  lineHeight: 1.56,
  letterSpacing: '0.1em',
  textTransform: 'uppercase',
  cursor: 'pointer',
  [CSS_SELECTOR_ACTIVE]: {
    transform: `translate(${buttons.clickOffset}, ${buttons.clickOffset})`
  }
});

const solidStyles = css(
  {
    backgroundColor: colours.primary,
    color: colours.white,
    [CSS_SELECTOR_HOVER]: {
      backgroundColor: colours.primaryDark
    },
    [CSS_SELECTOR_ACTIVE]: {
      backgroundColor: colours.primaryDark
    }
  },
  baseStyles
);

const hollowStyles = css(
  {
    backgroundColor: colours.transparent,
    boxShadow: `0 0 0 2px ${colours.primary} inset`,
    color: colours.primary,
    fontWeight: 'normal',
    [CSS_SELECTOR_HOVER]: {
      boxShadow: `0 0 0 2px ${colours.primaryDark} inset`,
      color: colours.primaryDark
    },
    [CSS_SELECTOR_ACTIVE]: {
      boxShadow: `0 0 0 2px ${colours.primaryDark} inset`,
      color: colours.primaryDark
    }
  },
  baseStyles
);

const styleMap = {
  [BUTTON_VARIANT_SOLID]: solidStyles,
  [BUTTON_VARIANT_HOLLOW]: hollowStyles
};

const Button = ({
  className,
  label,
  variant,
  cssOverrides,
  ...buttonProps
}) => (
  <div className={className}>
    <button
      type="button"
      {...buttonProps}
      css={[styleMap[variant], ...cssOverrides]}
    >
      {label}
    </button>
  </div>
);

Button.propTypes = {
  label: PropTypes.string,
  variant: PropTypes.oneOf([BUTTON_VARIANT_HOLLOW, BUTTON_VARIANT_SOLID]),
  className: PropTypes.string,
  cssOverrides: PropTypes.arrayOf(PropTypes.object)
};

Button.defaultProps = {
  /** The text for the button */
  label: 'button',
  /** The design variant */
  variant: BUTTON_VARIANT_SOLID,
  /** Class name for the button */
  className: null,
  /** Css overrides for the button */
  cssOverrides: []
};

export default Button;
