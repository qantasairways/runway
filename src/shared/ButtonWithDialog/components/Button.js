import React from 'react';
import PropTypes from 'prop-types';
import { css } from '@emotion/core';

import noop from '../../../utils/noop';

import {
  fontSize,
  colours,
  letterSpacing,
  layout,
  fontFamily,
  fontWeight
} from '../../../theme/airways';

function buttonStyles({ hasMediumGreyback }) {
  return css({
    label: 'runway-popup-field__button',
    position: 'relative',
    height: '70px',
    width: '100%',
    maxWidth: '100%',
    border: 0,
    padding: `0 ${layout.gutter}`,
    cursor: 'pointer',
    backgroundColor: hasMediumGreyback
      ? colours.mediumGrey
      : colours.darkerGrey,
    color: colours.white,
    fontFamily: fontFamily.main,
    fontSize: fontSize.body,
    fontWeight: fontWeight.regular
  });
}

export function ButtonContent({ largeButtonValue, smallButtonValue }) {
  return (
    <div
      css={{
        width: '100%',
        minWidth: 0
      }}
    >
      <div
        css={{
          label: 'runway-dialog-button__value--large',
          fontSize: fontSize.large,
          letterSpacing: letterSpacing.small,
          lineHeight: 1
        }}
      >
        {largeButtonValue}
      </div>
      <div
        css={{
          label: 'runway-dialog-button__value--small',
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          whiteSpace: 'nowrap'
        }}
      >
        {smallButtonValue}
      </div>
    </div>
  );
}

ButtonContent.propTypes = {
  largeButtonValue: PropTypes.string,
  smallButtonValue: PropTypes.string
};

ButtonContent.defaultProps = {
  largeButtonValue: '',
  smallButtonValue: ''
};

function Button({
  renderButtonValue,
  iconLabelButtonValue,
  onClick,
  onBlur,
  setButtonRef,
  open,
  buttonLabel,
  placeHolder,
  Icon
}) {
  const buttonValue = renderButtonValue();

  const renderIconLabelPair = () => (
    <div
      css={{
        width: '100%',
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center'
      }}
    >
      <span
        css={{
          label: 'runway-popup-field__label',
          flex: '1',
          whiteSpace: 'nowrap',
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          textAlign: 'left'
        }}
      >
        {iconLabelButtonValue.label}
      </span>
      <span
        css={{
          width: layout.iconSize,
          height: layout.iconSize
        }}
      >
        <iconLabelButtonValue.icon
          css={{
            label: 'runway-popup-field__icon',
            right: '10px',
            fill: colours.white
          }}
          height="100%"
          width="100%"
        />
      </span>
    </div>
  );

  return (
    <button
      type="button"
      aria-haspopup="dialog"
      aria-expanded={open}
      onClick={onClick}
      onBlur={onBlur}
      ref={setButtonRef}
      css={
        iconLabelButtonValue
          ? buttonStyles({ hasMediumGreyback: true })
          : buttonStyles({ hasMediumGreyback: false })
      }
    >
      {iconLabelButtonValue ? (
        renderIconLabelPair()
      ) : (
        <div
          css={{
            label: 'runway-popup-field__label',
            top: '5px',
            left: '7px',
            color: 'white',
            textTransform: 'none',
            fontSize: fontSize.label,
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center'
          }}
        >
          <span
            css={{
              label: 'runway-popup-field__label',
              position: 'absolute',
              top: '5px',
              left: '7px',
              color: 'white',
              fontSize: fontSize.label
            }}
          >
            {buttonLabel}
          </span>
          {buttonValue}
          {!buttonValue && (
            <div
              css={{
                label: 'runway-popup-field__placeholder',
                color: colours.mediumDarkGrey,
                fontSize: fontSize.large,
                letterSpacing: letterSpacing.small
              }}
            >
              {placeHolder}
            </div>
          )}
          {!buttonValue && !!Icon && (
            <Icon
              css={{
                label: 'runway-popup-field__icon',
                position: 'absolute',
                right: '10px',
                fill: colours.mediumDarkGrey
              }}
              height={layout.iconSize}
              width={layout.iconSize}
            />
          )}
        </div>
      )}
    </button>
  );
}

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  renderButtonValue: PropTypes.func,
  onBlur: PropTypes.func,
  setButtonRef: PropTypes.func,
  open: PropTypes.bool,
  buttonLabel: PropTypes.string,
  placeHolder: PropTypes.string,
  Icon: PropTypes.func,
  iconLabelButtonValue: PropTypes.shape({
    label: PropTypes.string,
    icon: PropTypes.any
  })
};

Button.defaultProps = {
  onBlur: noop,
  setButtonRef: noop,
  renderButtonValue: () => null,
  open: false,
  buttonLabel: '',
  placeHolder: '',
  Icon: null,
  iconLabelButtonValue: null
};

export default Button;
