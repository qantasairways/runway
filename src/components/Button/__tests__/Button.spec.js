import React from 'react';
import Button from '..';
import { mount } from 'enzyme';

describe('Button', () => {
  it('should render correctly in "debug" mode', () => {
    const component = mount(<Button debug="true" />);

    expect(component).toMatchSnapshot();
  });
});
