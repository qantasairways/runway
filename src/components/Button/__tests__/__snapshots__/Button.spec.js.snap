// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly in "debug" mode 1`] = `
.emotion-0 {
  background-color: #ee0000;
  color: #fff;
  font-family: "Ciutadella",sans-serif;
  font-size: 1rem;
  width: 100%;
  height: 3.125rem;
  border-radius: 4px;
  border-width: 0;
  padding: 14px 24px 15px;
  font-weight: 700;
  line-height: 1.56;
  -webkit-letter-spacing: 0.1em;
  -moz-letter-spacing: 0.1em;
  -ms-letter-spacing: 0.1em;
  letter-spacing: 0.1em;
  text-transform: uppercase;
  cursor: pointer;
}

.emotion-0:hover {
  background-color: #ba0000;
}

.emotion-0:active {
  background-color: #ba0000;
}

.emotion-0:active {
  -webkit-transform: translate(1px,1px);
  -ms-transform: translate(1px,1px);
  transform: translate(1px,1px);
}

<div
  className={null}
>
  <button
    className="emotion-0"
    debug="true"
    type="button"
  >
    button
  </button>
</div>
`;
